File: agent/web/services/model_service.py
from __future__ import annotations

import copy
import time
from typing import Any

from loguru import logger

from ...model_config import (
    build_provider_snapshot,
    load_model_config,
    mark_entry_vision,
    save_model_config,
)
from ...providers.registry import provider_options


# 2x2 JPEG probe used for vision connectivity tests.
_PROBE_JPEG_BASE64 = (
    "/9j/4AAQSkZJRgABAQAASABIAAD/4QBMRXhpZgAATU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAAqADAAQAAAABAAAAAgAAAAD/7QA4UGhvdG9zaG9wIDMuMAA4QklNBAQAAAAAAAA4QklNBCUAAAAAABDUHYzZjwCyBOmACZjs+EJ+/8AAEQgAAgACAwEiAAIRAQMRAf/EAB8AAAEFAQEBAQEBAAAAAAAAAAABAgMEBQYHCAkKC//EALUQAAIBAwMCBAMFBQQEAAABfQECAwAEEQUSITFBBhNRYQcicRQygZGhCCNCscEVUtHwJDNicoIJChYXGBkaJSYnKCkqNDU2Nzg5OkNERUZHSElKU1RVVldYWVpjZGVmZ2hpanN0dXZ3eHl6g4SFhoeIiYqSk5SVlpeYmZqio6Slpqeoqaqys7S1tre4ubrCw8TFxsfIycrS09TV1tfY2drh4uPk5ebn6Onq8fLz9PX29/j5+v/EAB8BAAMBAQEBAQEBAQEAAAAAAAABAgMEBQYHCAkKC//EALURAAIBAgQEAwQHBQQEAAECdwABAgMRBAUhMQYSQVEHYXETIjKBCBRCkaGxwQkjM1LwFWJy0QoWJDThJfEXGBkaJicoKSo1Njc4OTpDREVGR0hJSlNUVVZXWFlaY2RlZmdoaWpzdHV2d3h5eoKDhIWGh4iJipKTlJWWl5eYmZqio6Slpqeoqaqys7S1tre4ubrCw8TFxsfIycrS09TV1tfY2drh4uPk5ebn6Onq8fLz9PX29/j5+v/bAEMAAgICAgICAwICAwUDAwMFBgUFBQUGCAYGBgYGCAoICAgICAgKCgoKCgoKCgwMDAwMDA4ODg4ODw8PDw8PDw8PD//bAEMBAgICBAQEBwQEBxALCQsQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEP/dAAQAAf/aAAwDAQACEQMRAD8A/CQeNPGIAA12/AH/AE9S/wDxVL/wmvjL/oPX/wD4FS//ABVczRX3B/RB/9k="
)


class ModelService:
    def __init__(self, state) -> None:
        self.state = state
        self.root = state.root

    def payload(self) -> dict[str, Any]:
        config = load_model_config(self.root)
        from ...model_config import _resolve_active_entry

        entry = _resolve_active_entry(config, None)
        loop = self.state.loop
        secondary_snapshot = (
            loop.model_router.secondary
            if hasattr(loop, "model_router")
            else build_provider_snapshot(self.root, role="secondary")
        )
        return {
            "current": {
                "provider": loop.provider_name,
                "providerLabel": loop.provider_label,
                "model": loop.model,
                "apiBase": loop.provider_snapshot.api_base,
                "maxTokens": loop.max_tokens,
                "temperature": loop.temperature,
                "reasoningEffort": loop.reasoning_effort,
                "contextWindowTokens": loop.max_context,
                "entryName": entry.name,
                "entryLabel": entry.label or entry.name,
                "supportsVision": bool(getattr(loop, "supports_vision", False)),
                "mainModelId": entry.main_model_id,
                "secondaryModelId": entry.secondary_model_id,
                "modelRole": getattr(loop, "model_role", "main"),
            },
            "secondary": {
                "provider": secondary_snapshot.provider_name,
                "providerLabel": secondary_snapshot.provider_label,
                "model": secondary_snapshot.model,
                "apiBase": secondary_snapshot.api_base,
                "maxTokens": secondary_snapshot.generation.max_tokens,
                "temperature": secondary_snapshot.generation.temperature,
                "reasoningEffort": secondary_snapshot.generation.reasoning_effort,
                "contextWindowTokens": secondary_snapshot.context_window_tokens,
                "entryName": entry.name,
                "entryLabel": entry.label or entry.name,
                "supportsVision": bool(secondary_snapshot.supports_vision),
                "mainModelId": entry.main_model_id,
                "secondaryModelId": entry.secondary_model_id,
                "modelRole": secondary_snapshot.model_role,
            } if entry.secondary_model_id else None,
            "routing": loop.model_router.payload() if hasattr(loop, "model_router") else {
                "secondaryEnabled": bool(entry.secondary_model_id),
                "fallbackToMain": True,
            },
            "config": self.redact_apikeys(config.raw),
            "providerOptions": provider_options(),
        }

    def save(self, config: dict[str, Any]) -> None:
        existing = load_model_config(self.root).raw
        self._restore_masked_keys(config, existing)
        save_model_config(self.root, config, validate_complete=True)
        self.state.loop.refresh_model_config()

    async def test(self, body: dict[str, Any]) -> tuple[dict[str, Any], int]:
        entry_name = str(body.get("entryName") or "").strip()
        kind = str(body.get("kind") or "text").lower()
        role = str(body.get("role") or "main").lower()
        if kind not in {"text", "vision"}:
            return {"ok": False, "kind": kind, "error": "kind must be 'text' or 'vision'"}, 400
        if not entry_name:
            return {"ok": False, "kind": kind, "error": "entryName required"}, 400
        if role not in {"main", "secondary"}:
            return {"ok": False, "kind": kind, "error": "role must be 'main' or 'secondary'"}, 400
        if kind == "vision":
            role = "main"

        config = load_model_config(self.root)
        entry = config.find_entry(entry_name)
        if role == "secondary" and entry is not None and not entry.secondary_model_id:
            return {
                "ok": False,
                "kind": kind,
                "error": "secondaryModelId is required before testing the secondary model",
            }, 400

        try:
            snap = build_provider_snapshot(self.root, model_override=entry_name, role=role)
        except Exception as exc:
            return {"ok": False, "kind": kind, "error": f"snapshot failed: {exc}"}, 200

        messages = _vision_probe_messages() if kind == "vision" else [
            {"role": "user", "content": "Reply with exactly one word: pong"}
        ]
        started = time.monotonic()
        try:
            resp = await snap.provider.chat(
                messages=messages,
                tools=None,
                model=snap.model,
                max_tokens=64,
                temperature=0.0,
                reasoning_effort=None,
            )
        except Exception as exc:
            return {
                "ok": False,
                "kind": kind,
                "error": str(exc),
                "latencyMs": int((time.monotonic() - started) * 1000),
                "model": snap.model,
                "provider": snap.provider_name,
                "modelRole": snap.model_role,
            }, 200

        latency = int((time.monotonic() - started) * 1000)
        sample = (resp.content or "").strip()[:200]
        ok = _vision_ok(sample) if kind == "vision" else bool(sample) and "pong" in sample.lower()
        payload: dict[str, Any] = {
            "ok": ok,
            "kind": kind,
            "latencyMs": latency,
            "model": snap.model,
            "provider": snap.provider_name,
            "modelRole": snap.model_role,
            "sample": sample,
            "finishReason": getattr(resp, "finish_reason", "stop"),
        }
        if kind == "vision" and ok:
            try:
                mark_entry_vision(self.root, entry_name, value=True)
                self.state.loop.refresh_model_config()
                payload["visionMarked"] = True
            except Exception as exc:
                logger.warning(f"failed to mark entry vision: {exc}")
                payload["visionMarked"] = False
        return payload, 200

    @staticmethod
    def redact_apikeys(raw: dict) -> dict:
        out = copy.deepcopy(raw)

        def _mask(key: str) -> str:
            return "***" + key[-4:] if len(key) > 4 else "***"

        for prov in out.get("providers", {}).values():
            if isinstance(prov, dict) and isinstance(prov.get("apiKey"), str) and prov["apiKey"]:
                prov["apiKey"] = _mask(prov["apiKey"])
        for entry in out.get("models", []) or []:
            if isinstance(entry, dict) and isinstance(entry.get("apiKey"), str) and entry["apiKey"]:
                entry["apiKey"] = _mask(entry["apiKey"])
        return out

    @staticmethod
    def _restore_masked_keys(config: dict[str, Any], existing: dict[str, Any]) -> None:
        incoming_providers = config.get("providers") or {}
        existing_providers = existing.get("providers") or {}
        for name, prov in incoming_providers.items():
            if isinstance(prov, dict) and isinstance(prov.get("apiKey"), str) and prov["apiKey"].startswith("***"):
                prov["apiKey"] = existing_providers.get(name, {}).get("apiKey", "")

        incoming_models = config.get("models") or []
        existing_models = {
            m.get("name"): m for m in (existing.get("models") or [])
            if isinstance(m, dict) and m.get("name")
        }
        for entry in incoming_models:
            if not isinstance(entry, dict):
                continue
            api_key = entry.get("apiKey")
            if isinstance(api_key, str) and api_key.startswith("***"):
                entry["apiKey"] = existing_models.get(entry.get("name"), {}).get("apiKey", "")


def _vision_probe_messages() -> list[dict[str, Any]]:
    return [{
        "role": "user",
        "content": [
            {"type": "text", "text": "Reply with ONE English word only: name a visible color in this image."},
            {"type": "image_url", "image_url": {"url": f"data:image/jpeg;base64,{_PROBE_JPEG_BASE64}"}},
        ],
    }]


def _vision_ok(sample: str) -> bool:
    lowered = sample.lower()
    return bool(sample) and not any(
        token in lowered for token in ("invalid", "error", "cannot", "unable", "sorry")
    )

File: packages/core/src/api/services/model-service.ts
import { resolve } from 'node:path'
import {
  activateModelEntry,
  activeEntry,
  deleteModelEntry,
  findEntry,
  loadModelConfig,
  maskSecret,
  parseModelConfig,
  saveModelEntry,
  upsertModelEntryConfig,
  type ModelConfig,
  type ModelEntry,
  type ModelEntryUpdate,
  type ModelEntryV2,
  type ModelProtocol,
} from '../../config/model-config'
import { modelAvailability, type ModelAvailability } from '../../model/availability'
import { resolveModelProfile, type ResolvedModelProfile } from '../../model/profile'
import {
  buildProviderSnapshot,
  type ModelRoute,
  type ProviderSnapshot,
} from '../../model/router'
import type { OpenAiMessage } from '../../providers/base'
import {
  findByName,
  normalizeApiBase,
  providerOptions,
  type ProviderOption,
  type ProviderSpec,
} from '../../providers/registry'
import type { ProfileOnboardingActionResult } from '../../sessions/onboarding'

type Dict = Record<string, any>

const MODEL_DISCOVERY_TIMEOUT_MS = 15_000
const ANTHROPIC_MODELS_URL = 'https://api.anthropic.com/v1/models'
const ANTHROPIC_VERSION = '2023-06-01'

export interface CoreModelRouterLike {
  route(
    useCase: string,
    agentType?: string | null,
    task?: string | null,
  ): ModelRoute
  payload?(): Record<string, unknown>
}

export interface CoreModelServiceDeps {
  router: CoreModelRouterLike | (() => CoreModelRouterLike)
  refreshModelConfig?: () => void | Promise<void>
  afterConfigSaved?: () =>
    | ProfileOnboardingActionResult
    | Promise<ProfileOnboardingActionResult>
}

export type ModelEntrySaveInput = Omit<
  ModelEntryUpdate,
  'legacy'
>

export interface ModelEntryPayload
  extends Omit<ModelEntryV2, 'apiKey' | 'legacy'> {
  apiKey: string
  resolvedProfile: ResolvedModelProfile
}

export interface CurrentModelPayload {
  entryId: string
  provider: string
  providerLabel: string
  protocol: ModelProtocol
  modelId: string
  displayName: string | null
  apiBase: string
  reasoningEffort: string | null
  contextWindowTokens: number
  maxTokens: number
  capabilities: Pick<
    ResolvedModelProfile,
    'toolCall' | 'vision' | 'reasoning'
  >
  capabilitySources: ResolvedModelProfile['sources']
  reasoningEfforts: ResolvedModelProfile['reasoningEfforts']
  reasoningAdapter: ResolvedModelProfile['reasoningAdapter']
}

export interface ModelConfigPayload {
  schemaVersion: 2
  activeModelId: string | null
  models: ModelEntryPayload[]
  current: CurrentModelPayload | null
  availability: ModelAvailability
  providerOptions: ProviderOption[]
}

export interface ModelConfigSavePayload extends ModelConfigPayload {
  profileOnboarding?: ProfileOnboardingActionResult
}

export interface DiscoveredModel {
  id: string
  ownedBy?: string
  created?: number | string
}

export interface ModelDiscoveryPayload {
  ok: boolean
  provider: string
  protocol: ModelProtocol
  apiBase: string | null
  source: string
  models: DiscoveredModel[]
  code?: string
  message?: string
}

export class CoreModelService {
  readonly root: string
  private readonly deps: CoreModelServiceDeps
  private profileOnboardingRequested = false

  constructor(root: string, deps: CoreModelServiceDeps) {
    this.root = resolve(root)
    this.deps = deps
  }

  async getConfig(): Promise<ModelConfigPayload> {
    const config = await loadModelConfig(this.root)
    const entry = activeEntry(config) ?? null
    const models = config.raw.models.map((item) => modelEntryPayload(item))
    return {
      schemaVersion: 2,
      activeModelId: config.activeModelId,
      models,
      current: entry ? currentModelPayload(entry) : null,
      availability: modelAvailability(config),
      providerOptions: providerOptions(),
    }
  }

  async saveEntry(input: ModelEntrySaveInput): Promise<ModelConfigSavePayload> {
    const before = await loadModelConfig(this.root)
    const wasUsable = modelAvailability(before).usable
    const update = normalizeEntrySecret(input)
    await saveModelEntry(this.root, update)
    return this.afterModelMutation(wasUsable)
  }

  async deleteEntry(entryId: string): Promise<ModelConfigSavePayload> {
    const before = await loadModelConfig(this.root)
    await deleteModelEntry(this.root, requiredId(entryId))
    return this.afterModelMutation(modelAvailability(before).usable)
  }

  async activate(entryId: string): Promise<ModelConfigSavePayload> {
    const before = await loadModelConfig(this.root)
    await activateModelEntry(this.root, requiredId(entryId))
    return this.afterModelMutation(modelAvailability(before).usable)
  }

  async setReasoningEffort(
    entryId: string,
    reasoningEffort: string | null,
  ): Promise<ModelConfigPayload> {
    const config = await loadModelConfig(this.root)
    const entry = findEntry(config, requiredId(entryId))
    if (!entry) throw new Error(`model entry not found: ${entryId}`)
    const profile = resolvedProfile(entry)
    if (
      reasoningEffort !== null &&
      !profile.reasoningEfforts.includes(reasoningEffort as any)
    )
      throw new Error(
        `模型 ${entry.modelId} 不支持思考强度 ${reasoningEffort}`,
      )
    await saveModelEntry(this.root, {
      entryId: entry.entryId,
      reasoningEffort,
    })
    await this.deps.refreshModelConfig?.()
    return this.getConfig()
  }

  async discoverModels(input: Dict): Promise<ModelDiscoveryPayload> {
    const config = await loadModelConfig(this.root)
    const entryId = trimString(input.entryId)
    const entry = entryId ? (findEntry(config, entryId) ?? null) : null
    const providerName = trimString(input.provider) || entry?.provider || ''
    const spec = findByName(providerName)
    if (!spec) throw new Error(`provider 无效: ${providerName}`)
    const submittedProtocol = trimString(input.protocol)
    if (spec.name === 'custom' && !submittedProtocol)
      throw new Error('custom provider 必须明确选择 protocol')
    const protocol = (submittedProtocol || entry?.protocol || spec.defaultProtocol) as
      | ModelProtocol
      | null
    if (!protocol || !spec.protocols.includes(protocol))
      throw new Error(`provider ${spec.name} 不支持 protocol: ${protocol ?? ''}`)

    const submittedBase = trimString(input.apiBase)
    const apiBase = normalizeApiBase(
      protocol,
      submittedBase || entry?.apiBase || spec.apiBases[protocol] || '',
    )
    const canReuseEntryCredentials = discoveryIdentityMatches(
      entry,
      spec,
      protocol,
      apiBase,
    )
    const apiKey = discoveryApiKey(
      input.apiKey,
      entry,
      canReuseEntryCredentials,
    )
    const extraHeaders = discoveryExtraHeaders(
      input.extraHeaders,
      entry,
      canReuseEntryCredentials,
    )
    const discovery = spec.modelDiscovery[protocol] ?? 'unsupported'

    if (discovery === 'unsupported')
      return discoveryUnavailable(
        spec,
        protocol,
        apiBase,
        'unsupported_protocol',
        `${spec.displayName} 当前不支持通过 ${protocol} 自动获取模型列表，请手动填写模型 ID。`,
      )
    if (!spec.isLocal && !spec.isOauth && !apiKey)
      return discoveryUnavailable(
        spec,
        protocol,
        apiBase,
        'credential_required',
        `请先填写 ${spec.displayName} 的 API Key 后再获取模型列表。`,
      )
    if (!apiBase)
      return discoveryUnavailable(
        spec,
        protocol,
        apiBase,
        'missing_api_base',
        `请先填写 ${spec.displayName} 的 API Base 后再获取模型列表。`,
      )

    return discovery === 'anthropic'
      ? discoverAnthropicModels(spec, protocol, apiBase, apiKey, extraHeaders)
      : discoverOpenAiCompatibleModels(
          spec,
          protocol,
          apiBase,
          apiKey,
          extraHeaders,
        )
  }

  async test(body: { entryId?: string; kind?: string }): Promise<Dict> {
    const entryId = trimString(body.entryId)
    const kind = trimString(body.kind || 'text').toLowerCase()
    if (kind !== 'text' && kind !== 'vision')
      return { ok: false, kind, error: "kind must be 'text' or 'vision'" }
    if (!entryId) return { ok: false, kind, error: 'entryId required' }

    const config = await loadModelConfig(this.root)
    const entry = findEntry(config, entryId)
    if (!entry) return { ok: false, kind, error: `model entry not found: ${entryId}` }

    let snapshot: ProviderSnapshot
    try {
      snapshot = this.snapshotForModelTest(config, entryId, kind === 'vision')
    } catch (error) {
      return {
        ok: false,
        kind,
        error: `snapshot failed: ${error instanceof Error ? error.message : String(error)}`,
      }
    }

    const started = Date.now()
    try {
      const response = await snapshot.provider.chat({
        messages:
          kind === 'vision'
            ? visionProbeMessages()
            : [{ role: 'user', content: 'Reply with exactly one word: pong' }],
        tools: null,
        model: snapshot.model,
        maxTokens: 64,
        temperature: 0,
        reasoningEffort: null,
      })
      const sample = String(response.content || '').trim().slice(0, 200)
      const ok = kind === 'vision' ? visionOk(sample) : /pong/i.test(sample)
      const payload: Dict = {
        ok,
        kind,
        entryId,
        latencyMs: Date.now() - started,
        model: snapshot.model,
        provider: snapshot.providerName,
        sample,
        finishReason: response.finishReason || 'stop',
      }
      return payload
    } catch (error) {
      return {
        ok: false,
        kind,
        entryId,
        error: error instanceof Error ? error.message : String(error),
        latencyMs: Date.now() - started,
        model: snapshot.model,
        provider: snapshot.providerName,
      }
    }
  }

  private snapshotForModelTest(
    config: ModelConfig,
    entryId: string,
    forceVision = false,
  ): ProviderSnapshot {
    if (!forceVision) return buildProviderSnapshot(config, { modelOverride: entryId })
    const entry = findEntry(config, entryId)
    if (!entry) throw new Error(`model entry not found: ${entryId}`)
    const raw = upsertModelEntryConfig(config.raw, {
      entryId,
      capabilityOverrides: { ...entry.capabilityOverrides, vision: true },
    })
    return buildProviderSnapshot(parseModelConfig(raw), { modelOverride: entryId })
  }

  private async afterModelMutation(
    wasUsable: boolean,
  ): Promise<ModelConfigSavePayload> {
    await this.deps.refreshModelConfig?.()
    const payload = await this.getConfig()
    if (
      !wasUsable &&
      payload.availability.usable &&
      !this.profileOnboardingRequested
    ) {
      const profileOnboarding = await this.deps.afterConfigSaved?.()
      if (profileOnboarding) {
        this.profileOnboardingRequested = true
        return { ...payload, profileOnboarding }
      }
    }
    return payload
  }
}

function requiredId(value: string): string {
  const id = value.trim()
  if (!id) throw new Error('entryId required')
  return id
}

function normalizeEntrySecret(input: ModelEntrySaveInput): ModelEntryUpdate {
  const update = structuredClone(input) as ModelEntryUpdate
  if (
    typeof update.apiKey === 'string' &&
    (!update.apiKey.trim() || update.apiKey.trim().startsWith('***'))
  )
    delete update.apiKey
  return update
}

function resolvedProfile(entry: ModelEntry | ModelEntryV2): ResolvedModelProfile {
  const modelId =
    entry.modelId || ('mainModelId' in entry ? entry.mainModelId : '')
  return resolveModelProfile({
    provider: entry.provider,
    protocol: entry.protocol ?? 'openai',
    modelId,
    capabilityOverrides: entry.capabilityOverrides,
    contextWindowTokens: entry.contextWindowTokens ?? undefined,
    maxTokens: entry.maxTokens ?? undefined,
  })
}

function modelEntryPayload(entry: ModelEntryV2): ModelEntryPayload {
  const { legacy: _legacy, ...safe } = entry
  return {
    ...safe,
    apiKey: maskSecret(entry.apiKey),
    resolvedProfile: resolvedProfile(entry),
  }
}

function currentModelPayload(entry: ModelEntry): CurrentModelPayload {
  const profile = resolvedProfile(entry)
  const spec = findByName(entry.provider)
  return {
    entryId: entry.entryId || entry.name,
    provider: entry.provider,
    providerLabel: spec?.displayName ?? entry.provider,
    protocol: entry.protocol ?? spec?.defaultProtocol ?? 'openai',
    modelId: entry.modelId || entry.mainModelId,
    displayName: entry.displayName || entry.label || null,
    apiBase: entry.apiBase || '',
    reasoningEffort: entry.reasoningEffort,
    contextWindowTokens: profile.contextWindowTokens,
    maxTokens: profile.maxTokens,
    capabilities: {
      toolCall: profile.toolCall,
      vision: profile.vision,
      reasoning: profile.reasoning,
    },
    capabilitySources: profile.sources,
    reasoningEfforts: profile.reasoningEfforts,
    reasoningAdapter: profile.reasoningAdapter,
  }
}

function discoveryIdentityMatches(
  entry: ModelEntry | null,
  spec: ProviderSpec,
  protocol: ModelProtocol,
  apiBase: string,
): boolean {
  if (!entry || entry.provider !== spec.name) return false
  const entryProtocol = entry.protocol ?? spec.defaultProtocol
  if (entryProtocol !== protocol) return false
  const entryApiBase = normalizeApiBase(
    protocol,
    entry.apiBase || spec.apiBases[protocol] || '',
  )
  return entryApiBase === apiBase
}

function discoveryApiKey(
  input: unknown,
  entry: ModelEntry | null,
  canReuseEntryCredentials: boolean,
): string {
  if (input === null) return ''
  const direct = trimString(input)
  const existing = trimString(entry?.apiKey)
  if (!direct || direct.startsWith('***'))
    return canReuseEntryCredentials ? existing : ''
  return direct
}

function discoveryExtraHeaders(
  input: unknown,
  entry: ModelEntry | null,
  canReuseEntryCredentials: boolean,
): Record<string, string> {
  const resolved =
    recordValue(input) ||
    (input !== null && canReuseEntryCredentials
      ? recordValue(entry?.legacy?.extraHeaders)
      : null) ||
    {}
  return Object.fromEntries(
    Object.entries(resolved)
      .filter(([, value]) => value !== null && value !== undefined)
      .map(([key, value]) => [key, String(value)]),
  )
}

async function discoverOpenAiCompatibleModels(
  spec: ProviderSpec,
  protocol: ModelProtocol,
  apiBase: string | null,
  apiKey: string,
  extraHeaders: Record<string, string>,
): Promise<ModelDiscoveryPayload> {
  let last: ModelDiscoveryPayload | null = null
  for (const url of openAiModelEndpointCandidates(apiBase || '')) {
    try {
      const response = await fetchWithTimeout(url, {
        method: 'GET',
        headers: discoveryHeaders(
          extraHeaders,
          apiKey ? { authorization: `Bearer ${apiKey}` } : {},
        ),
      })
      if (!response.ok) {
        last = discoveryUnavailable(
          spec,
          protocol,
          apiBase,
          codeForHttpStatus(response.status),
          response.statusText || `HTTP ${response.status}`,
        )
        continue
      }
      return {
        ok: true,
        provider: spec.name,
        protocol,
        apiBase,
        source: 'openai_compat',
        models: parseDiscoveredModels(await response.json()),
      }
    } catch (error) {
      last = discoveryUnavailable(
        spec,
        protocol,
        apiBase,
        codeForFetchError(error),
        fetchErrorMessage(error),
      )
    }
  }
  return (
    last ??
    discoveryUnavailable(
      spec,
      protocol,
      apiBase,
      'no_endpoint',
      '未找到可用的模型列表 endpoint。',
    )
  )
}

async function discoverAnthropicModels(
  spec: ProviderSpec,
  protocol: ModelProtocol,
  apiBase: string | null,
  apiKey: string,
  extraHeaders: Record<string, string>,
): Promise<ModelDiscoveryPayload> {
  try {
    const response = await fetchWithTimeout(anthropicModelsUrl(apiBase), {
      method: 'GET',
      headers: discoveryHeaders(extraHeaders, {
        'x-api-key': apiKey,
        'anthropic-version': ANTHROPIC_VERSION,
      }),
    })
    if (!response.ok)
      return discoveryUnavailable(
        spec,
        protocol,
        apiBase,
        codeForHttpStatus(response.status),
        response.statusText || `HTTP ${response.status}`,
      )
    return {
      ok: true,
      provider: spec.name,
      protocol,
      apiBase,
      source: 'anthropic',
      models: parseDiscoveredModels(await response.json()),
    }
  } catch (error) {
    return discoveryUnavailable(
      spec,
      protocol,
      apiBase,
      codeForFetchError(error),
      fetchErrorMessage(error),
    )
  }
}

function discoveryUnavailable(
  spec: ProviderSpec,
  protocol: ModelProtocol,
  apiBase: string | null,
  code: string,
  message: string,
): ModelDiscoveryPayload {
  return {
    ok: false,
    provider: spec.name,
    protocol,
    apiBase,
    source: spec.modelDiscovery[protocol] ?? 'unsupported',
    models: [],
    code,
    message,
  }
}

async function fetchWithTimeout(url: string, init: RequestInit): Promise<Response> {
  const controller = new AbortController()
  const timer = setTimeout(() => controller.abort(), MODEL_DISCOVERY_TIMEOUT_MS)
  try {
    return await fetch(url, { ...init, signal: controller.signal })
  } finally {
    clearTimeout(timer)
  }
}

function discoveryHeaders(
  extraHeaders: Record<string, string>,
  defaults: Record<string, string>,
): Headers {
  const headers = new Headers({ accept: 'application/json', ...defaults })
  for (const [key, value] of Object.entries(extraHeaders)) headers.set(key, value)
  return headers
}

function openAiModelEndpointCandidates(apiBase: string): string[] {
  const base = trimApiBase(apiBase)
  return base ? [`${base}/models`] : []
}

function anthropicModelsUrl(apiBase: string | null): string {
  const base = trimApiBase(apiBase || '')
  if (!base) return ANTHROPIC_MODELS_URL
  if (base.endsWith('/v1')) return `${base}/models`
  return `${base}/v1/models`
}

function parseDiscoveredModels(body: unknown): DiscoveredModel[] {
  const data = modelArray(body)
  const seen = new Set<string>()
  const out: DiscoveredModel[] = []
  for (const raw of data) {
    if (!raw || typeof raw !== 'object' || Array.isArray(raw)) continue
    const item = raw as Record<string, unknown>
    const id = trimString(item.id) || trimString(item.name) || trimString(item.model)
    if (!id || seen.has(id)) continue
    seen.add(id)
    const model: DiscoveredModel = { id }
    const ownedBy = trimString(item.owned_by) || trimString(item.ownedBy) || trimString(item.owner)
    const created = item.created ?? item.created_at ?? item.createdAt
    if (ownedBy) model.ownedBy = ownedBy
    if (typeof created === 'number' || typeof created === 'string') model.created = created
    out.push(model)
  }
  return out.sort((a, b) => a.id.localeCompare(b.id))
}

function modelArray(body: unknown): unknown[] {
  if (Array.isArray(body)) return body
  if (!body || typeof body !== 'object' || Array.isArray(body)) return []
  const record = body as Record<string, unknown>
  if (Array.isArray(record.data)) return record.data
  if (Array.isArray(record.models)) return record.models
  return []
}

function codeForHttpStatus(status: number): string {
  if (status === 401 || status === 403) return 'auth'
  if (status === 429) return 'rate_limit'
  if (status >= 500 || status === 408) return 'transient'
  return 'request_failed'
}

function codeForFetchError(error: unknown): string {
  return error instanceof Error && error.name === 'AbortError'
    ? 'timeout'
    : 'network_error'
}

function fetchErrorMessage(error: unknown): string {
  if (error instanceof Error && error.name === 'AbortError') return '获取模型列表超时。'
  return error instanceof Error ? error.message : String(error)
}

function visionProbeMessages(): OpenAiMessage[] {
  return [
    {
      role: 'user',
      content: [
        { type: 'text', text: 'What is the color of this image? Reply with one word.' },
        {
          type: 'image_url',
          image_url: {
            // 1×1 opaque red RGBA PNG. Keep the expected answer aligned with visionOk().
            url: 'data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAYAAAAfFcSJAAAADUlEQVR4nGP4z8DwHwAFAAH/iZk9HQAAAABJRU5ErkJggg==',
          },
        },
      ],
    },
  ]
}

function visionOk(sample: string): boolean {
  const normalized = sample.trim().toLowerCase().replace(/[.!。！]+$/g, '')
  return normalized === 'red' || normalized === '红' || normalized === '红色'
}

function trimString(value: unknown): string {
  return typeof value === 'string' ? value.trim() : ''
}

function trimApiBase(value: string): string {
  return value.trim().replace(/\/+$/, '')
}

function recordValue(value: unknown): Record<string, unknown> | null {
  return value && typeof value === 'object' && !Array.isArray(value)
    ? (value as Record<string, unknown>)
    : null
}
